app/services/course/duplication/object_duplication_service.rb

```diff
@@ -40,17 +40,46 @@ def time_shift(current_course, target_course)
   # @return [Object] The duplicated object, if `objects` is a single object.
   # @return [Array] Array of duplicated objects, if `objects` is an array.
   def duplicate_objects(objects)
-    # TODO: Inform the user when the duplication is complete.
+    # TODO: Email the user when the duplication is complete.
     Course.transaction do
       duplicated = duplicator.duplicate(objects)
-      success = duplicated.respond_to?(:save) ? duplicated.save : duplicated.all?(&:save)
-      raise ActiveRecord::Rollback unless success
+      save_success = duplicated.respond_to?(:save) ? duplicated.save : duplicated.all?(&:save)
+      after_save_success = save_success && after_save(objects, duplicated)
+      raise ActiveRecord::Rollback unless after_save_success
       duplicated
     end
   end
 
   private
 
+  # Executes callbacks meant to be invoked after duplicated objects have been saved
+  #
+  # @param [Object|Array] _objects The source object(s)
+  # @param [Object|Array] duplicated The duplicated object(s)
+  # @return [Boolean] true if all callbacks are executed successfully
+  def after_save(_objects, duplicated)
+    populate_assessment_category_default_tabs(duplicated)
+  end
+
+  # Creates the mandatory tab for duplicated assessment categories if no tabs are already present.
+  # This cannot be done when duplicating the category since we do not know if a tab belonging to it will
```

Comment:
I would prefer no component specific code here, but just for discussion on potentially better ways:

1) Should we consider some duplication 'rules' (defined at component level)? For one, I think it doesn't make sense to duplicate a category without a tab. 
2) The other way that we can implement the `after_save` as a DSL for components to declare callbacks.

Reply:
Fixed. 

The course duplication service [has some `Category` specific code also](https://github.com/Coursemology/coursemology2/pull/2383#pullrequestreview-52574278). We did this because it was ugly to make it work with the save-the-course-and-everything-saves-automatically approach. Now that we are creating and saving the course first, agree that it might be good to move updating of category email settings from the course duplication service to the category model also. Can do the next time we work on the course duplicator.

The currently implemented 'rule' for category duplication is:
- If a category's child tab has been duplicated during the current job, just establish the relation
- If none of a category's child tabs has been duplicated, don't create a tab first. Wait until after after all items have been duplicated to see if any of its child tabs are duplicated afterward. If not, create a default tab.